docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+As described in [Issue#591](https://github.com/gardener/etcd-druid/issues/591) scheduling compaction only based on number of accumulated etcd events is not sufficient to ensure a successful compaction. This is specifically targeted for kubernetes clusters where each etcd event is larger in size owing to large spec or status fields or respective resources. 
+
+Druid will now need information regarding snapshot sizes, and more importantly the total size of accumulated delta snapshots since the last full snapshot.
+
+The authors propose to enhance the proposed `snapshots` field described in [Use Case #3](#use-case-3-monitoring-backup-health) with the following additional field:
+
+```yaml
+snapshots:
+  accumulatedDeltaSize: <total size of delta snapshots since last full snapshot>
+```
+
+Druid can then use this information in addition to the existing revision information to decide to trigger an early snapshot compaction job. This effectively allows druid to be proactive in performing regular compactions for etcds receiving large events, reducing the probability of a failed snapshot compaction or restoration.
+
+### Enhanced Defragmentation
+
+Reader is recommended to read [Etcd Compaction & Defragmentation](https://etcd.io/docs/v3.4/op-guide/maintenance/) in order to understand the following terminology:
+
+[`dbSize`](https://github.com/etcd-io/etcd/blob/a603c0798948941d453b158af794edab1a8230be/mvcc/backend/backend.go#L60-L64) - total storage space used by the etcd database
+
+[`dbSizeInUse`](https://github.com/etcd-io/etcd/blob/a603c0798948941d453b158af794edab1a8230be/mvcc/backend/backend.go#L65-L68) - logical storage space used by the etcd database, not accounting for free pages in the DB due to etcd history compaction
+
+The leading-backup-sidecar performs periodic defragmentations of the DBs of all the etcd-members in the cluster, controlled via a defragmentation cron schedule provided to each backup-sidecar. Defragmentation is a costly maintenance operation and causes a brief downtime to the etcd-member being defragmented, due to which the leading-backup-sidecar defragments each etcd-member sequentially. This ensures that only one etcd-member would be unavailable at any given time, thus avoiding an accidental quorum loss in the etcd cluster.
+
+The authors propose to move the responsibility of orchestrating these individual defragmentations to druid due to the following reasons:
+
+- Since each backup-sidecar only has knowledge of the health of its own etcd, it can only determine whether its own etcd can be defragmented or not, based on etcd-member health. Trying to defragment a different healthy etcd-member while another etcd-member is unhealthy would lead to a transient quorum loss.
```

Comment:
> Since each backup-sidecar only has knowledge of the health of its own etcd, it can only determine whether its own etcd can be defragmented or not, based on etcd-member health

That's absolutely wrong assumption, in actual leading sidecar backup-restore has cluster wide information about health of all etcd cluster members, please check this code: 
https://github.com/gardener/etcd-backup-restore/blob/5bb95a80183d7b31c284cfcc4b9c0d5bdb50c581/pkg/defragmentor/defrag.go#L81


> Trying to defragment a different healthy etcd-member while another etcd-member is unhealthy would lead to a transient quorum loss.

sorry, I didn't get this part. We already thought through this kind-of scenario while working on multi-node etcd design.  please check this [doc of leader_election](https://github.com/gardener/etcd-backup-restore/blob/master/doc/concepts/leader_election.md#defragmentation)
We only trigger defragmentation when all of the members of etcd cluster returns `true` of `etcd status API call` i.e all members are healthy else leading sidecar backup-restore keep checking the health of all members of etcd cluster until all etcd cluster members become healthy.

Reply:
The question is not whether leading backup sidecar currently has knowledge of the entire cluster, but whether it should, going forward. If leading backup "sidecar" has knowledge of the entire cluster, then it is no longer a sidecar but an orchestrator, and it interferes with the responsibilities of the actual cluster orchestrator, aka, etcd-druid.

> We only trigger defragmentation when all of the members of etcd cluster returns true of etcd status API call i.e all members are healthy else leading sidecar backup-restore keep checking the health of all members of etcd cluster until all etcd cluster members become healthy

Got it, thanks. So this means that the leading backup sidecar simply waits until the cluster is fully healthy before it can trigger a defragmentation. But the operator who can likely recover the cluster from an unhealthy state is still druid, so it doesn't do any harm if druid is the one who initiates a cluster-wide defragmentation operation, again because druid should be the one who has overall knowledge of all the members, not any one backup sidecar.

Reply:
In addition, druid can pause the entire defragmentation operation if there is a transient quorum loss. This can happen if lets say it is currently issued a defragmentation request to etcd-member-1. While etcd-member-1 is running defragmentation, etcd-member-2 goes down and does not come up for some time. Now we have a quorum loss. Druid will know this and pause any further defragmentation. Even after the current defragmentation is over and quorum is restored, if the etcd-member-2 still is down then it will not continue. Once all members are up, then it can resume from where it left (provided we are still far away from the scheduled defragmentation schedule). This is a responsibility of an orchestrator/operator and not of a side-car.

Reply:
>  If leading backup "sidecar" has knowledge of the entire cluster, then it is no longer a sidecar but an orchestrator, and it interferes with the responsibilities of the actual cluster orchestrator, aka, etcd-druid.

yes, I agree that `backup-restore` is now no longer just a sidecar as it used to be, may be in initial days of backup-restore. May be we need to redefine some of our terminologies as we are just keep using the old terminologies.
Why I'm saying this because `backup-restore` also provides the configuration to corresponding etcd, by sidecar logic this feature of `providing config` should also change ... right ?
IMO, we shouldn't label backup-restore as sidecar now and we should redefine/revisit functionalities performed by etcd-backup-restore, so it become easy to decide which way we want to go. WDYT ?

Reply:
>Why I'm saying this because backup-restore also provides the configuration to corresponding etcd, by sidecar logic this feature of providing config should also change ... right ?

And this will also change. It will return to just being a sidecar which provides backup and restore functionality.

Reply:
>And this will also change. It will return to just being a sidecar which provides backup and restore functionality and anything around it.

This will not be addressed as part of this proposal but there will be a separate proposal which details out how the responsibilities will change between etcd-wrapper and backup-restore containers.

Reply:
thanks @unmarshall,
I'm just letting this comment thread open for future references.